### ISSUE TITLE
Wallet not accruately tracking utxos

### ISSUE THREAD

**Issue by Christewart** [bitcoin-s/bitcoin-s#3868]
My wallet ended up not tracking states correctly. I'm unsure of why this is occurring at the moment. I believe this corruption was happening slowly over time, but i need to look into the specific utxos that are different from the two files. One reason the utxos sets are different is that the `wallet-utxos.txt` contained reserved utxos that are not included in the [`getutxos`](https://github.com/bitcoin-s/bitcoin-s/blob/31e83245228d207f9c5a3f2d230c9397c6fe1e34/wallet/src/main/scala/org/bitcoins/wallet/internal/UtxoHandling.scala#L41) response. I'm attaching two files to this issue, the first file `wallet-utxos.txt` is the result of `getutxos` before a rescan.

The second file `wallet-utxos-2.txt` is the rest of `getutxos` after a rescan.

Balance of the wallet before rescan

```
 ./bitcoin-s-cli getbalances
{
  "confirmed": 0.04385147,
  "unconfirmed": 1.4195E-4,
  "reserved": 0.0859621,
  "total": 0.12995552
}
```

Balance after rescan

```
 ./bitcoin-s-cli getbalances
{
  "confirmed": 1.33076847,
  "unconfirmed": 0,
  "reserved": 0,
  "total": 1.33076847
}
```
The contents of `wallet-utxos.txt`

```
[
  {
    "outpoint": {
      "txid": "c4698bf9afad0714f3cd8704a154985e3193cbfcc8cdaf754a806cacabed7fd8",
      "vout": 1
    },
    "value": 1179
  },
  {
    "outpoint": {
      "txid": "8c137c7f5000e5ce0035f9fb868acf7cc5706891a9da17d0e44e3033dd15c0aa",
      "vout": 91
    },
    "value": 547
  },
  {
    "outpoint": {
      "txid": "8c137c7f5000e5ce0035f9fb868acf7cc5706891a9da17d0e44e3033dd15c0aa",
      "vout": 105
    },
    "value": 547
  },
  {
    "outpoint": {
      "txid": "9974b0465681e3f77a02da23e8ef0a2b0703faf5ca248e5389b9d85afd41ff1f",
      "vout": 0
    },
    "value": 36435
  },
  {
    "outpoint": {
      "txid": "3a12f1154b8394ecc72b2cb7c9ad7d8a4110d7051cbc48955c22914c8d23e3c0",
      "vout": 0
    },
    "value": 3035
  },
  {
    "outpoint": {
      "txid": "52a1ee395661345b2fde4761ecf1c9035bbc449ed87ac7038eb6fb8123340ff1",
      "vout": 2
    },
    "value": 10667
  },
  {
    "outpoint": {
      "txid": "f44571102b28cfbac046dc73ec75c699a9c4767df58e7ebbfc6e1f1f4052600e",
      "vout": 0
    },
    "value": 5953
  },
  {
    "outpoint": {
      "txid": "d9586c8492444620dab565998b75510d17164f1e487423dd3d664bedc4ff82df",
      "vout": 0
    },
    "value": 2831
  },
  {
    "outpoint": {
      "txid": "bbe6d5d670d48d8353dc7e561a3c0a0bdc9424a3848547f30b2c21367d5c53c2",
      "vout": 0
    },
    "value": 12358
  },
  {
    "outpoint": {
      "txid": "0d1e8c2572fe5483b53fb711243c4d11893b58b64af759f439c22c5fb34c0f05",
      "vout": 0
    },
    "value": 14195
  },
  {
    "outpoint": {
      "txid": "c23e3c9982f3cd1776c9a1269ed5d858324d2e41bddb32b152c97d7d30527ae0",
      "vout": 0
    },
    "value": 500000
  },
  {
    "outpoint": {
      "txid": "c23e3c9982f3cd1776c9a1269ed5d858324d2e41bddb32b152c97d7d30527ae0",
      "vout": 1
    },
    "value": 3811595
  }
]

```

The contents of `wallet-utxos-2.txt`

```
[
  {
    "outpoint": {
      "txid": "9ca2f15bd4032445761ca3a6a215ca7b70383d650dc0643cf7792f3a127baa06",
      "vout": 0
    },
    "value": 1000000
  },
  {
    "outpoint": {
      "txid": "a58c48724bcf9fca74f3a2dbf35016a6bd605d421ac4809b95c19008bb0d9b29",
      "vout": 0
    },
    "value": 405000
  },
  {
    "outpoint": {
      "txid": "ab44fdb5c9e8ed664c5fca1680ac4448c66eb5f8aafe1a7563b5ed8d01a7197f",
      "vout": 0
    },
    "value": 100000
  },
  {
    "outpoint": {
      "txid": "7f8e4e754c9145a7516c9226c3266624efdd2d4eca446d6965b1cb1778652584",
      "vout": 1
    },
    "value": 182000
  },
  {
    "outpoint": {
      "txid": "9d9f93be1c16173e4107a9acbd9ca7da3375c9c4a71ba0403b561f92dbc86b50",
      "vout": 0
    },
    "value": 40000
  },
  {
    "outpoint": {
      "txid": "3c6be4bdc33f5fbac6a5ff74a2a5a71bb6ae8d5acbc1ec0f9f8876b4852a42a7",
      "vout": 0
    },
    "value": 100000
  },
  {
    "outpoint": {
      "txid": "db2366864933dcb50226b39cbd69002f32bb324496a78a5ee83f17db87e9ea29",
      "vout": 1
    },
    "value": 14939450
  },
  {
    "outpoint": {
      "txid": "eea2fa05edda7b269973ab35de891881952bf418d8c48d27ced77aaa0c8bf9ff",
      "vout": 1
    },
    "value": 18075077
  },
  {
    "outpoint": {
      "txid": "3e5e34dcfa619f4795fd2dcbf1a0d119bb9fdbd05e6652e3cd10c4c81bb05347",
      "vout": 0
    },
    "value": 100000
  },
  {
    "outpoint": {
      "txid": "9229dae78bb9f0d49cd4058d6ea0067ec0c1d0264141dabda3ffd04ca539a462",
      "vout": 1
    },
    "value": 35738964
  },
  {
    "outpoint": {
      "txid": "e9f9d96cfe04a7238f46baebc7c19f6596a18164ca22e2a19dcdc634fbc71553",
      "vout": 0
    },
    "value": 100000
  },
  {
    "outpoint": {
      "txid": "0f5b834f8b4ab642d32d13beea08057df64d3e749cb1635436791257b60ea035",
      "vout": 0
    },
    "value": 100000
  },
  {
    "outpoint": {
      "txid": "9974b0465681e3f77a02da23e8ef0a2b0703faf5ca248e5389b9d85afd41ff1f",
      "vout": 0
    },
    "value": 36435
  },
  {
    "outpoint": {
      "txid": "b7084948b97165e26d5df4a3e5bfc60c2274a821bfe54b18c190ce1fd9f6544b",
      "vout": 2
    },
    "value": 8531779
  },
  {
    "outpoint": {
      "txid": "fafb64a619444d69d2229c4c4cbbcd32461af46d4902c9129b7b701871a2b044",
      "vout": 1
    },
    "value": 50000
  },
  {
    "outpoint": {
      "txid": "4d54cd2a193656b1ec08d998c45d13d8f02a4f7671ad5c97154546a6ce138445",
      "vout": 0
    },
    "value": 3041947
  },
  {
    "outpoint": {
      "txid": "0e0148d0e1d8af57c35d104b52f6405aeae1ff99f5acf0e1727dc76bd934fe36",
      "vout": 0
    },
    "value": 18180
  },
  {
    "outpoint": {
      "txid": "de13c464b401ef130c7c67b2fa456b127f4461f0978957d63a12959c3cb6e98d",
      "vout": 1
    },
    "value": 33580
  },
  {
    "outpoint": {
      "txid": "be73ff50dcb2926b9840c249541612f59bf6130577ddd7ca58d41f8398709778",
      "vout": 0
    },
    "value": 46172840
  },
  {
    "outpoint": {
      "txid": "c23e3c9982f3cd1776c9a1269ed5d858324d2e41bddb32b152c97d7d30527ae0",
      "vout": 0
    },
    "value": 500000
  },
  {
    "outpoint": {
      "txid": "c23e3c9982f3cd1776c9a1269ed5d858324d2e41bddb32b152c97d7d30527ae0",
      "vout": 1
    },
    "value": 3811595
  }
]
```

[wallet-utxos.txt](https://github.com/bitcoin-s/bitcoin-s/files/7644551/wallet-utxos.txt)

[wallet-utxos-2.txt](https://github.com/bitcoin-s/bitcoin-s/files/7645746/wallet-utxos-2.txt)

**Comment by Christewart** [comment]
So to summarize it seems that my wallet has none of those dust outputs after the rescan :thinking: . It also seems the rescan found more funds that were missed by the wallet during normal operations.

I've used this wallet with both `bitcoin-s.node.mode={bitcoind,neutrino}`, so i'm unsure what backend wasn't providing correct block data. The first wallet does have some valid utxos that weren't found during rescan it seems, for instance [8c137c7f5000e5ce0035f9fb868acf7cc5706891a9da17d0e44e3033dd15c0aa](https://mempool.space/tx/8c137c7f5000e5ce0035f9fb868acf7cc5706891a9da17d0e44e3033dd15c0aa)

**Comment by Christewart** [comment]
## Valid outputs that are in the first wallet but not in the second rescanned wallet

[8c137c7f5000e5ce0035f9fb868acf7cc5706891a9da17d0e44e3033dd15c0aa:91](https://mempool.space/tx/8c137c7f5000e5ce0035f9fb868acf7cc5706891a9da17d0e44e3033dd15c0aa)

[8c137c7f5000e5ce0035f9fb868acf7cc5706891a9da17d0e44e3033dd15c0aa:105](8c137c7f5000e5ce0035f9fb868acf7cc5706891a9da17d0e44e3033dd15c0aa)



## Valid outputs that are in the rescanned wallet, but not in the first wallet

[9ca2f15bd4032445761ca3a6a215ca7b70383d650dc0643cf7792f3a127baa06:0](https://mempool.space/tx/9ca2f15bd4032445761ca3a6a215ca7b70383d650dc0643cf7792f3a127baa06)

[a58c48724bcf9fca74f3a2dbf35016a6bd605d421ac4809b95c19008bb0d9b29:0](https://mempool.space/tx/a58c48724bcf9fca74f3a2dbf35016a6bd605d421ac4809b95c19008bb0d9b29)

[ab44fdb5c9e8ed664c5fca1680ac4448c66eb5f8aafe1a7563b5ed8d01a7197f:0](https://mempool.space/tx/ab44fdb5c9e8ed664c5fca1680ac4448c66eb5f8aafe1a7563b5ed8d01a7197f)

[7f8e4e754c9145a7516c9226c3266624efdd2d4eca446d6965b1cb1778652584:1](https://mempool.space/tx/7f8e4e754c9145a7516c9226c3266624efdd2d4eca446d6965b1cb1778652584)

[9d9f93be1c16173e4107a9acbd9ca7da3375c9c4a71ba0403b561f92dbc86b50:0](https://mempool.space/tx/9d9f93be1c16173e4107a9acbd9ca7da3375c9c4a71ba0403b561f92dbc86b50)

[3c6be4bdc33f5fbac6a5ff74a2a5a71bb6ae8d5acbc1ec0f9f8876b4852a42a7:0](https://mempool.space/tx/3c6be4bdc33f5fbac6a5ff74a2a5a71bb6ae8d5acbc1ec0f9f8876b4852a42a7)

[db2366864933dcb50226b39cbd69002f32bb324496a78a5ee83f17db87e9ea29:1](https://mempool.space/tx/db2366864933dcb50226b39cbd69002f32bb324496a78a5ee83f17db87e9ea29)

[eea2fa05edda7b269973ab35de891881952bf418d8c48d27ced77aaa0c8bf9ff:1](https://mempool.space/tx/eea2fa05edda7b269973ab35de891881952bf418d8c48d27ced77aaa0c8bf9ff)

[3e5e34dcfa619f4795fd2dcbf1a0d119bb9fdbd05e6652e3cd10c4c81bb05347:0](https://mempool.space/tx/3e5e34dcfa619f4795fd2dcbf1a0d119bb9fdbd05e6652e3cd10c4c81bb05347)

[9229dae78bb9f0d49cd4058d6ea0067ec0c1d0264141dabda3ffd04ca539a462:1](https://mempool.space/tx/9229dae78bb9f0d49cd4058d6ea0067ec0c1d0264141dabda3ffd04ca539a462)

[e9f9d96cfe04a7238f46baebc7c19f6596a18164ca22e2a19dcdc634fbc71553:0](https://mempool.space/tx/e9f9d96cfe04a7238f46baebc7c19f6596a18164ca22e2a19dcdc634fbc71553)

[0f5b834f8b4ab642d32d13beea08057df64d3e749cb1635436791257b60ea035:0](https://mempool.space/tx/0f5b834f8b4ab642d32d13beea08057df64d3e749cb1635436791257b60ea035)

[b7084948b97165e26d5df4a3e5bfc60c2274a821bfe54b18c190ce1fd9f6544b:2](https://mempool.space/tx/b7084948b97165e26d5df4a3e5bfc60c2274a821bfe54b18c190ce1fd9f6544b)

[fafb64a619444d69d2229c4c4cbbcd32461af46d4902c9129b7b701871a2b044:1](https://mempool.space/tx/fafb64a619444d69d2229c4c4cbbcd32461af46d4902c9129b7b701871a2b044)

[4d54cd2a193656b1ec08d998c45d13d8f02a4f7671ad5c97154546a6ce138445:0](https://mempool.space/tx/4d54cd2a193656b1ec08d998c45d13d8f02a4f7671ad5c97154546a6ce138445)

[0e0148d0e1d8af57c35d104b52f6405aeae1ff99f5acf0e1727dc76bd934fe36:0](https://mempool.space/tx/0e0148d0e1d8af57c35d104b52f6405aeae1ff99f5acf0e1727dc76bd934fe36)

[de13c464b401ef130c7c67b2fa456b127f4461f0978957d63a12959c3cb6e98d:1](https://mempool.space/tx/de13c464b401ef130c7c67b2fa456b127f4461f0978957d63a12959c3cb6e98d)

[be73ff50dcb2926b9840c249541612f59bf6130577ddd7ca58d41f8398709778:0](https://mempool.space/tx/be73ff50dcb2926b9840c249541612f59bf6130577ddd7ca58d41f8398709778)

[]()

**Comment by Christewart** [comment]
fixed in #4290 / #4296
